Add support for "freestyle" in multiplayer

smoogipoo opened the pull request:
- [x] https://github.com/ppy/osu-web/pull/11760
- [x] https://github.com/ppy/osu-server-spectator/pull/251
- Closes https://github.com/ppy/osu/issues/5578

|         | Old Server           | New Server  |
| ------------- |:-------------:| :-----:|
| Old Client      | 🟢  | 🟢  |
| New Client      | 🟢  | 🟢 |

## Preamble

I was told to effectively hack this in, and was not given an opportunity to refactor multiplayer to properly support it or even to be able to write tests without pulling my hair out. So here goes.

Have I done some basic testing to see things work? Yes.
Will things break? Definitely.
Do I know what will break? Nope.

## Description

https://github.com/user-attachments/assets/723d9cc0-7266-4c30-8fec-57b6116a6c5c

I've named this "freestyle", because I see a future where this is more than just beatmap/ruleset but also mods and thus include "freemods" in it. Perhaps something like a menu where you can choose what is free.

For the time being, when it's enabled, it will disable mods/freemods. I'm not sure what the intended UX is, but it will definitely be a lot more complicated due to mod types not matching between rulesets, i.e. this will fail a ton of validation everywhere. So I'm keeping it simple.

There's no display in the room for what/how _other_ users have chosen to play, I'm not even sure where I'd add that in because the room design is a whole bunch of mess. There's already no display of other users' freemods either, so /shrug?

smoogipoo committed:
Adjust test to load song select during setup

smoogipoo committed:
Add BeatmapSetId to playlist items

smoogipoo committed:
Add "freeplay" button to multiplayer song select

smoogipoo committed:
Fix free mod button overriding enabled state

smoogipoo committed:
Fix incorrect behaviour

smoogipoo committed:
Initial support for free style selection

smoogipoo committed:
Only allow selecting beatmaps within 30s length

smoogipoo committed:
Block beatmap presents in style select screen

smoogipoo committed:
Adjust namings

smoogipoo committed:
Add style selection to playlists screen

smoogipoo committed:
Fix freemods button opening overlay unexpectedly

smoogipoo committed:
Add some comments

smoogipoo committed:
Use online ruleset_id to build local score models

smoogipoo committed:
Fix inspection

peppy commented:
> There's already no display of other users' freemods either, so /shrug?

![2024-12-24 21 08 35@2x](https://github.com/user-attachments/assets/c5e1a927-287e-4511-a7cd-a5932792d1fc)

It's shown here. I think adding the ruleset icon in the same flow area at minimum would be good to have.

peppy commented:
@cl8n able to provide some basic design direction for how ruleset and difficulty selection could be shown per-user here?

smoogipoo committed:
Fix co-variant array conversion

smoogipoo committed:
Fix test

smoogipoo committed:
Rewrite everything to better support spectator server messaging

smoogipoo committed:
Display user style in participant panel

smoogipoo committed:
Don't display on matching beatmap/ruleset

smoogipoo commented:
Had to pretty much rewrite the entire screen and add server-side support to do the above ¯\\_(ツ)_/¯. Added it as a ruleset icon displayed on the panel:

https://github.com/user-attachments/assets/fe4a50e9-39a0-4ccc-8119-7c684fc06cd2

smoogipoo committed:
Remove unused usings

smoogipoo committed:
Change to using a 'FreeStyle' boolean

smoogipoo committed:
Fix possible null reference

smoogipoo committed:
Fix style display refreshing on all room updates

smoogipoo committed:
Send `beatmap_id` when creating score

peppy committed:
Merge branch 'master' into multiplayer-free-style

peppy committed:
Add tooltips explaining multiplayer mod selection buttons

peppy committed:
Perform client side validation that the selected beatmap and ruleset have valid online IDs

This is local to playlists, since in multiplayer the validation is
already provided by `osu-server-spectator`.

peppy commented:
I've done an initial pass and fixed a few things up. Overall, the code is bearable (there are changes I'd propose for code quality reasons, but it's isolated and at this point it wouldn't be achieving too much – as you know the whole multiplayer screens / drawable structuring isn't great).

Remaining requirements to push this live in my eyes are:

- Showing freestyle status on lobby listings (filtering to rooms with it enabled?)
  - We might want to contact a designer with spare time (icon designer chan maybe) to get something for this 
  - Bonus: also showing freemod when it's enabled
- Forcing freemod to enabled when freestyle is on, and adding ability to choose base mods
  - In IRL discussion, the proposal is to do this in a follow-up PR to reduce complexity. I'm okay with that as we have a couple of weeks before the next release cycle.

I'll do a second pass tomorrow with the goal of getting this merged, unless there are any blockers.

peppy reviewed (commented):
As above

smoogipoo commented:
> Showing freestyle status on lobby listings (filtering to rooms with it enabled?)

I believe this will need additional osu-web support because we only receive the current playlist item in the lounge, not the full playlist. Unless that's what you mean - for the badge to be shown only when the current item is freestyle?

peppy commented:
> I believe this will need additional osu-web support because we only receive the current playlist item in the lounge, not the full playlist. Unless that's what you mean - for the badge to be shown only when the current item is freestyle?

Current item should be fine for now.

peppy committed:
Merge branch 'master' into multiplayer-free-style

peppy commented:
I can't test the results screen because I can't turn on no fail, so I'm just going to have to assume it works and displays good. I'll continue with a code-only review and we can do further testing after this is pushed out to staging.

peppy commented:
@smoogipoo what causes the server breakage mentioned in the OP? I've had a look and can't find anything that would break things completely.

peppy commented on osu.Game/Online/Multiplayer/MultiplayerRoomUser.cs:
```diff
@@ -37,6 +34,21 @@ public class MultiplayerRoomUser : IEquatable<MultiplayerRoomUser>
+        public MatchUserState? MatchState { get; set; }
+
+        /// <summary>
+        /// Any ruleset applicable only to the local user.
```
"Any" doesn't read great here. How about

> If not-null, a local override for this user's ruleset selection.

> If not-null, a local override for this user's beatmap selection.

peppy committed:
Combine countless update methods all called together into a single method

smoogipoo committed:
Adjust xmldocs

smoogipoo committed:
Add pill displaying current freestyle status

peppy committed:
Combine more methods to simplify flow

peppy commented:
I'm going to do some refactoring because the following is very confusing:

- `GetGameplayRuleset` is `virtual` (and overridden by the multiplayer screen), but also reads from `UserRuleset` which is expected to be set by the multiplayer screen.
- `UserRuleset` sits alongside the existing `UserMods`, but `UserMods` is never written to outside `RoomSubScreen` while `UserRuleset` *is*.

peppy committed:
Add missing `partial`

peppy committed:
Remove new bindables from `RoomSubScreen`

peppy committed:
Fix delay before difficulty panel displays fully

peppy committed:
Combine more methods to simplify flow futher

peppy committed:
Minimal clean-up to get things bearable

I plan to do a full refactor of `RoomSubScreen` at first opportunity.

peppy commented:
@smoogipoo i'm still working through this, but can you have a check on the commits above? i don't think there should be anything too contested. some of the flow has changed / removed, but in testing it seems to still hold up so far.

smoogipoo commented:
I don't know if you've tested the whole `onRoomUpdated()` flow, because I know there's a reason it was done that way, but at face values the changes look reasonable.

peppy committed:
Standardise naming to use "Freestyle" not "FreeStyle"

peppy committed:
Tidy up button construction code

peppy commented:
I think things here are in a state that we could consider merging this, but mod support is not yet added.

@smoogipoo is this something you are interested in looking at, or would you prefer I take that on?

Either way, I think merging these base PRs is going to make testing and forward development easier. If we don't have the mods setup ready for the next release, we can just hide the footer UI button temporarily.

peppy committed:
Rename freestyle select screen classes for better discoverability

smoogipoo commented:
Your latest changes broke tests. Please make sure to fix those.

As I have mentioned in the past, I highly prefer to tackle everything in multiplayer myself.

peppy committed:
Remove `Scheduler.AddOnce` from `updateSpecifics`

To keep things simple, let's not bother debouncing this. The debouncing
was causing spectating handling to fail because of two interdependent
components binding to `BeatmapAvailability`:

Binding to update the screen's `Beatmap` after a download completes:

https://github.com/ppy/osu/blob/58747061171c4ebe70201dfe4d3329ed7f4343f5/osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs#L266-L267

Binding to attempt a load request:

https://github.com/ppy/osu/blob/8bb7bea04e56fab9247baa59ae879e16c8b4bd9b/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSubScreen.cs#L67

The first must update the beatmap before the second runs, else gameplay
will not load due to `Beatmap.IsDefault`.

peppy commented:
Roger, so to confirm you'll be giving priority to getting freemod working when freestyle is on to?

Fixed the test, see the commit message for details.

smoogipoo commented:
If you consider it to be of utmost importance, then yes, I can look at that as my immediate focus. But it will not be done in this PR.

peppy commented:
> If you consider it to be of utmost importance

The rough goal was for the next release (this week) but failing that it can happen the release after.

peppy committed:
Merge branch 'master' into multiplayer-free-style

smoogipoo commented:
This PR has been in review for over a month, and that hasn't been the blocker given we had an [IRL discussion about it](https://github.com/ppy/osu/pull/31260#issuecomment-2604393232). As I've made clear in the past - the quicker you merge my PRs, the faster we can come up with trailing work tasks, and the better I can focus on one specific area.

In the meantime, and in order not to conflict with potential changes here, I've divested my attention to other things that needed addressing - room interop flows, friend states, and user panel status.

Things will need to change if priorities have not been adequately assigned thus far.
